src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/handlers/handler_probe.h"
+
+#include "config/configuration.h"
+#include "kafka/server/handlers/handler_interface.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/metrics.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace kafka {
+
+handler_probe_manager::handler_probe_manager()
+  : _probes(max_api_key() + 2)
+  , _unknown_handler_index(max_api_key() + 1) {
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
+        }
+    }
+}
+
+handler_probe& handler_probe_manager::get_probe(api_key key) {
+    if (!handler_for_key(key)) {
+        key = api_key{_unknown_handler_index};
+    }
+
+    return *_probes.at(key);
+}
+
+handler_probe::handler_probe(api_key key)
+  : _key(key)
+  , _last_recorded_in_progress(ss::lowres_clock::now()) {}
+
+void handler_probe::setup_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    const char* handler_name;
+    if (auto handler = handler_for_key(_key)) {
+        handler_name = handler.value()->name();
+    } else {
+        handler_name = "unknown_handler";
+    }
+
+    std::vector<sm::label_instance> labels{sm::label("handler")(handler_name)};
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("kafka_handler"),
+      {
+        sm::make_counter(
+          "requests_completed_total",
+          [this] { return _requests_completed; },
+          sm::description("Number of kafka requests completed per handler"),
```

Comment:
As Ben mentioned, it might be best to leave off "per handler" since it is also "per" many other things like "shard", "instance" etc so handler is not necessarily special.

Reply:
Removed "per handler" from the description. Note there will be similar charts in the auto-generated dashboard for all of these metrics except they won't be tagged by handler. They can be found in `src/v/net/probes.cc`.

In the case of the one you commented on there will be a chart named `Kafka: Number of successful requests` in the auto-generated dashboard that is not tagged by handler. And there will be this chart `Number of kafka requests completed` that is tagged by handler.

Reply:
It's honestly a bit of a pain trying to name these. A good description of the metric doesn't make a for a good title for the auto-generate chart. I.e, if I wanted to describe the metric I'd use `Number of kafka requests completed`. But if I wanted to name the auto-generate chart that has the rate function applied to it I'd use `Kafka requests completed per second`.